ops/charm.py

```diff
@@ -9,17 +9,27 @@ class HookEvent(EventBase):
 
 class ActionEvent(EventBase):
 
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._params = None
+
     def defer(self):
         raise RuntimeError('cannot defer action events')
 
     def restore(self, snapshot):
+        self._params = None
         env_action_name = os.environ.get('JUJU_ACTION_NAME')
         event_action_name = self.handle.kind[:-len('_action')].replace('_', '-')
         if event_action_name != env_action_name:
             # This could only happen if the dev manually emits the action, or from a bug.
             raise RuntimeError('action event kind does not match current action')
-        # Params are loaded at restore rather than __init__ because the model is not available in __init__.
-        self.params = self.framework.model._backend.action_get()
+
+    @property
+    def params(self):
+        # NoneType will be returned if an action has no parameters defined but in that case this property is unlikely to be used.
```

Comment:
I think this is just "None will be returned".

However, if you call action-get, and we parse the result as JSON, don't we always end up with an Dict here, it will just be an empty dict if there are no parameters?

If we don't like returning None here, we can always do:
```
if self._params is None:
  self._params = ...
  if self._params is None:
    # no actual params, but avoid calling action-get again
    self._params = {}
return self._params
```

Reply:
I added some test code into one of my charms (locally) before submitting and validated that it's not an empty dict.

It's a bit difficult to debug without `debug-hooks` support for actions but I got this by inserting an rpdb trace and running an action:

```
(Pdb) subprocess.check_output(['action-get', '--format=json'])
b'null\n'

(Pdb) type(json.loads(subprocess.check_output(['action-get', '--format=json']).decode('utf-8')))
<class 'NoneType'>
```

I think I'll go for your suggestion with an empty dict.

Reply:
Interestingly, not using `--format=json` results in an empty dict being returned.

(Pdb) subprocess.check_output(['action-get']).decode('utf-8')                     
'{}\n'

Looks like a bug to me.

Reply:
Filed here: https://bugs.launchpad.net/juju/+bug/1862026

Reply:
Addressed in https://github.com/canonical/operator/pull/132/commits/61364bbf52b400d1a385f034046a4a84f7cdbf76